modules/renter/hostdb/hostdb.go

```diff
@@ -115,3 +119,22 @@ func New(cs modules.ConsensusSet, wallet modules.Wallet, tpool modules.Transacti
 
 	return hdb, nil
 }
+
+// newHostDB creates a HostDB using the provided dependencies. Unlike New, it
+// does not have any side effects (i.e. it does not spawn background threads,
+// perform I/O, or call stateful methods of its dependencies.)
+func newHostDB(w hdbWallet, tp hdbTransactionPool, d hdbDialer, s hdbSleeper, p hdbPersister, l hdbLogger) *HostDB {
+	return &HostDB{
+		wallet:  w,
+		tpool:   tp,
+		dialer:  d,
+		sleeper: s,
+		persist: p,
+		log:     l,
+
+		contracts:   make(map[types.FileContractID]hostContract),
+		activeHosts: make(map[modules.NetAddress]*hostNode),
+		allHosts:    make(map[modules.NetAddress]*hostEntry),
+		scanPool:    make(chan *hostEntry, scanPoolSize),
+	}
+}
```

Comment:
Originally I liked the idea of having no side effects because it made testing easier; no extra goroutines flying around or files written to disk. Another big advantage is that you can pass `nil` for most (or all) of the dependencies if you know your test isn't going to call them.

Now I'm reconsidering. If you test with just `newHostDB`, you could argue that you're not *really* testing a hostDB, since a real hostDB has goroutines running in the background.

Perhaps the better approach is to make `New` the tiny function and `newHostDB` the big one. e.g.:
```go
func New(cs hdbConsensusSet, wallet hdbWalletShim, tpool hdbTransactionPool,
    persistDir string) (*hostDB, error) {
    return newHostDB(cs, wallet, tpool, stdDialer{}, stdSleeper{}, stdPersister{},
        stdLogger{}, persistDir)
}

func newHostDB(cs hdbConsensusSet, w hdbWallet, tp hdbTransactionPool, d hdbDialer,
    s hdbSleeper, p hdbPersister, l hdbLogger) (*hostDB, error) {
   // .. initialize and call dependencies as needed ...
    return hdb, nil
}
```
This is kind of a bummer because you can no longer pass `nil` for the `os` dependency, since `newHostDB` will now call `MkdirAll`. But that seems like a small price to pay.